Uploading a Json Document which contains a null JsonObject Behavior

https://issues.couchbase.com/browse/CBSE-3779


Request
Should you be able to upload a JSON Document with null JSON Object as the content from an SDK, Java in this case?
Background
Poor programming skills found myself accidentally uploading a JSON Document with no content from the Java SDK to test out a Sync Gateway Webhook issue. When Sync Gateway pulled down the doc from a shadowed bucket Sync Gateway crashed.
Problem
From the Java SDK if a document is uploaded with,
bucket.async().upsert(JsonDocument.create("AAA_null", null))
the SDK will write this to the Couchbase Cluster,
ay 26, 2017 3:55:36 PM com.couchbase.client.deps.io.netty.handler.logging.LoggingHandler logMessage
FINEST: [id: 0x7bcdbf85, L:/10.112.151.1:55696 - R:node3-cb451-centos7.vagrants/10.112.151.103:11210] WRITE: 40B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 80 01 00 08 08 00 00 a1 00 00 00 14 00 00 00 02 |................|
|00000010| 00 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00 |................|
|00000020| 41 41 41 5f 6e 75 6c 6c                         |AAA_null        |
+--------+-------------------------------------------------+----------------+
May 26, 2017 3:55:36 PM com.couchbase.client.deps.io.netty.handler.logging.LoggingHandler logMessage
FINEST: [id: 0x7bcdbf85, L:/10.112.151.1:55696 - R:node3-cb451-centos7.vagrants/10.112.151.103:11210] WRITE: 4B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 6e 75 6c 6c                                     |null            |
+--------+-------------------------------------------------+----------------+
May 26, 2017 3:55:36 PM com.couchbase.client.deps.io.netty.handler.logging.LoggingHandler flush
I don't see any errors in Couchbase. If I go to the bucket and look at the documents, for this document I see,
ID             Content
AAA_null 	{{r.doc.json || r.doc.base64 | mnLimitTo: 200}}
If I edit the document I get the following warning "Warning: JSON object cannot be null" and the body of the document shows 'null'
I have this bucket shadowed in Sync Gateway. When Sync Gateway pulls the document down Sync Gateway will crash with the following stack,

2017-05-26T22:55:36.355Z Shadow+: Pulling "AAA_null", CAS=14c249f37add0000 ... have UpstreamRev="", UpstreamCAS=0
panic: assignment to entry in nil map
 
goroutine 77 [running]:
panic(0xade320, 0xc420bf56a0)
	/usr/local/go/1.7.4/go/src/runtime/panic.go:500 +0x1a1
github.com/couchbase/sync_gateway/db.(*Shadower).pullDocument.func1(0xc420514870, 0x4115b3, 0xc420bf5658, 0xc420509288, 0x8)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/shadower.go:128 +0x3b5
github.com/couchbase/sync_gateway/db.(*Database).updateDoc.func1(0xc420514870, 0x0, 0x0, 0x0, 0x0, 0xc420514870, 0x0)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/crud.go:660 +0x128
github.com/couchbase/sync_gateway/db.(*Database).updateDoc.func3(0x0, 0x0, 0x0, 0xba2d15, 0x7, 0xc4200fdb00, 0xa, 0x0, 0x0)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/crud.go:885 +0xee
github.com/couchbase/sync_gateway/base.CouchbaseBucketGoCB.WriteUpdate(0xc42007f170, 0xc4201980f0, 0x28, 0xba2d15, 0x7, 0xc4200fdb00, 0xa, 0x0, 0x0, 0xf5d980, ...)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:1116 +0x17e
github.com/couchbase/sync_gateway/base.(*CouchbaseBucketGoCBGoCouchbaseHybrid).WriteUpdate(0xc4201b22b0, 0xc420bf5518, 0x8, 0x0, 0xc420c51590, 0xc42024d6c0, 0xc4200baea8)
	<autogenerated>:553 +0xbb
github.com/couchbase/sync_gateway/db.(*Database).updateDoc(0xc420868aa0, 0xc420bf5518, 0x8, 0x0, 0xc420c51560, 0x0, 0x0, 0xc420bf5518, 0xc420bf5518)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/crud.go:894 +0x5a8
github.com/couchbase/sync_gateway/db.(*Shadower).pullDocument(0xc420110000, 0xc420bf5518, 0x8, 0xc420c51527, 0x4, 0x4, 0xc4202fb400, 0x14c249f37add0000, 0xc400000000, 0x0, ...)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/shadower.go:146 +0x35c
github.com/couchbase/sync_gateway/db.(*Shadower).readTapFeed(0xc420110000)
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/shadower.go:80 +0x332
created by github.com/couchbase/sync_gateway/db.NewShadower
	/home/couchbase/jenkins/workspace/sgw-unix-build@2/1.5.0/enterprise/godeps/src/github.com/couchbase/sync_gateway/db/shadower.go:39 +0x1a6
So should the SDK allow upserting documents with null content, should Couchbase allow it? If so I guess Sync Gateway should check for nil since this ends up crashing the process.
Logs
I can supply the client SDK, but other then the snippet I provided there is no error of any kind.
The Sync Gateway logs don't show anything other then the snippet provided pulling the document, then the stack.
Couchbase server logs show no errors.

Quick side note: "null" in JSON is a valid JSON document.

--- a/db/crud.go
+++ b/db/crud.go
@@ -329,7 +329,7 @@ func (db *DatabaseContext) getRevision(doc *document, revid string) (Body, error
 			return nil, base.HTTPErrorf(404, "missing")
 		} else if data, err := db.getOldRevisionJSON(doc.ID, revid); data == nil {
 			return nil, err
-		} else if err = json.Unmarshal(data, &body); err != nil {
+		} else if err = body.Unmarshal(data); err != nil {
 			return nil, err
 		}
 	}
@@ -598,7 +598,7 @@ func (db *Database) ImportRawDoc(docid string, value []byte, isDelete bool) (doc
 	if isDelete {
 		body = Body{"_deleted": true}
 	} else {
-		err := json.Unmarshal(value, &body)
+		err := body.Unmarshal(value)
 		if err != nil {
 			base.LogTo("Import", "Unmarshal error during importDoc %v", err)
 			return nil, err

--- a/db/revision.go
+++ b/db/revision.go
@@ -12,6 +12,7 @@ package db
 import (
 	"crypto/md5"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"net/http"
 	"strconv"
@@ -22,6 +23,17 @@ import (
 
 // The body of a CouchDB document/revision as decoded from JSON.
 type Body map[string]interface{}
+
+func (b *Body) Unmarshal(data []byte) error {
+
+	if err := json.Unmarshal(data, &b); err != nil {
+		return err
+	}
+	if b == nil {
+		return errors.New("Unable to unmarshal null document as Body")
+	}
+	return nil
+}
 
 func (body Body) ShallowCopy() Body {
 	copied := make(Body, len(body))

--- a/db/shadower.go
+++ b/db/shadower.go
@@ -1,7 +1,6 @@
 package db
 
 import (
-	"encoding/json"
 	"net/http"
 	"reflect"
 	"regexp"
@@ -97,7 +96,7 @@ func (s *Shadower) pullDocument(key string, value []byte, isDeletion bool, cas u
 	if isDeletion {
 		body = Body{"_deleted": true}
 	} else {
-		if err := json.Unmarshal(value, &body); err != nil {
+		if err := body.Unmarshal(value); err != nil {
 			base.LogTo("Shadow", "Doc %q is not JSON; skipping", key)
 			return nil
 		}

